### ISSUE TITLE
Lint/UselessAssignment causes a false positive when using hash value omission and block

### ISSUE THREAD

**Issue by sanfrecce-osaka** [rubocop/rubocop#13041]
Lint/UselessAssignment causes a false positive when using hash value omission and block.

## Expected behavior

The following code should be passed:

```ruby
def foo(bar)
  bar => { baz: }
  3.times { baz -= 1 }
  puts 'qux' if baz == -3
end
```

## Actual behavior

```ruby
def foo(bar)
  bar => { baz: }
  3.times { baz -= 1 } # 👮‍♂️ W: [Correctable] Lint/UselessAssignment: Useless assignment to variable - baz. Use - instead of -=.
  puts 'qux' if baz == -3
end
```

In contrast, the following codes pass.

```ruby
def foo(bar)
  baz = bar[:baz]
  3.times { baz -= 1 }
  puts 'qux' if baz == -3
end
```

```ruby
def foo(bar)
  bar => { baz: }
  baz -= 1 until baz == -3
  puts 'qux' if baz == -3
end
```

## RuboCop version

```
$ bundle exec rubocop -V
1.65.0 (using Parser 3.3.4.0, rubocop-ast 1.31.3, running on ruby 3.2.3) +server [arm64-darwin21]
  - rubocop-capybara 2.21.0
  - rubocop-factory_bot 2.26.1
  - rubocop-performance 1.21.1
  - rubocop-rails 2.25.1
  - rubocop-rspec 2.30.0
```

**Comment by koic** [comment]
It looks like a false negative happens even when the hash value is explicitly given, not just when it's omitted.
